### ISSUE TITLE
Support for minimum language version enforcement in macro deps

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/language#3466]
Spawned from a discord thread [here](https://discord.com/channels/608014603317936148/1169724112537657466/1172876172409962548).

Macros generate code which has some language version requirement, but the way current SDK constraints work, they are not sufficient to enforce that macros are only applied to packages that meet that requirement.

Consider the following scenario:

- There exists package `a`, which uses a macro from package `b`. Package `a` has a min SDK constraint of `3.0.0`, but package `b` does not generate code that uses any features from `>3.0.0` so everything is good.
- Package `b` then releases a new version, which requires a feature from `3.1.0`, but it is only used as an implementation detail. They update their min sdk constraint, and ship a feature release.
- Now, a user on the `3.1.0` (or greater) SDK runs a `pub upgrade`, which selects the new version of package `b` and they start getting errors (in package `a` 😭). This is because the macro re-runs on the same version of package `a`, but it now requires a language version of `3.1.0`, and it is running on libraries at version `3.0.0`.

All the SDK constraints are met here, but the existence of a package with an older language version also using that macro causes things to break down.

I think we will need to get a bit creative here, with a solution that is integrated into the version solver itself. In the scenario above, everything would have been fine if we just didn't select the newest version of the macro.

Today, any time a macro uses a new language feature, it would need to release that as a breaking change.

cc @jonasfj @sigurdm @natebosch @davidmorgan

**Comment by jakemac53** [comment]
Note that per-library language versions can also cause problems here. Using a macro in a library which has had its version downgraded via language version comment is probably generally unsafe, if that language version is below the min SDK constraint of the macro package.

**Comment by lrhn** [comment]
A macro should probably have a way to access the language version of the library is generative code for, so it can choose which solution to use for a given library, or it can fail gracefully or informatively of it cannot generate coffee for that language version.

There is nothing wrong with a macro refusing to work for a particular library, you just shoulder be using it then.

The problem comes if the macro library changes which library versions or works for. That should probably be a breaking change and require a major version increment for the macro.

Which language version a library supports can also affect the introspection API, and definitely the code generation API. Which may need to be versioned.

Macros are really contravariant in language version, since they don't (only) consume code, they produce it.

The min-SDK needed by a macro package is the SDK needed to run it.
It can have a different constraint on which language version library they can be applied to, and probably will.

**Comment by rrousselGit** [comment]
I don't think it is macros' responsibility to emit an error if they are run on a library with an incompatible language version.

For me, that's 100% the job of `pub get`. 
IMO it makes no sense for `pub get` to pass, only to end up with an error saying "incorrect language version, please upgrade".

**Comment by rrousselGit** [comment]
In any case, I don't think macro authors should have to consider this scenario.
Whether the error is triggered by `pub get` or when running the macro, IMO that error should be coming directly from the SDK.

This scenario is too intuitive for folks to consider it when working on their macros.
I've made the mistake a few times when working with build_runner, without even realizing it, and only realized what was causing the problem after years of working on code generators.

**Comment by jakemac53** [comment]
> A macro should probably have a way to access the language version of the library is generative code for, so it can choose which solution to use for a given library, or it can fail gracefully or informatively of it cannot generate coffee for that language version.

They can, but as @rrousselGit says this would be much better solved by the version solver.

> There is nothing wrong with a macro refusing to work for a particular library, you just shoulder be using it then.

The big problem here is that a macro might fail on some transitive dependency, not your own code.

> The problem comes if the macro library changes which library versions or works for. That should probably be a breaking change and require a major version increment for the macro.

Breaking changes are one option, but that could cause a lot of churn in macro package versions that could be expressed in a way that does not cause that churn. Especially for highly used macros, major version releases are very painful.

> Which language version a library supports can also affect the introspection API, and definitely the code generation API. Which may need to be versioned.

This is true but it matters less - it affects the package being actively developed and not transitive packages. It is definitely possible that I might not be able to use some new language feature immediately if I am using a macro that can't handle it, but I can just not use that feature until the macro does support it.

However when a macro starts using a new feature itself, that macro is no longer usable from any package on an earlier language version, so it can break any of your transitive deps. The solution would be to restrict the macro version, but you might not even have a dependency at all on that macro, so it both feels wrong to have to add that dependency just to enforce an older version, and it is probably much harder to figure out that is what you need to do.

> Macros are really contravariant in language version, since they don't (only) consume code, they produce it.
> It can have a different constraint on which language version library they can be applied to, and probably will.

This is exactly the point of this issue, discussing if we can have this 2nd constraint expressed in such a way that it can be a part of the pub solve :).

**Comment by rrousselGit** [comment]
I also doubt that many macros will use different constraints in their implementation than in the generated code. At least that wasn't my experience.

This is in part caused by lints.  
For example, when constructor tear-offs were introduced, the linter complained both in the implementation of my code-generators _and_ their outputs. So I naively changed both the generator and generated code at once ... and released it as a patch or minor release, like I would've done for any non-generator package.

I'm sure that mistake is going to be quite common. Even if the consequences aren't that bad

**Comment by jakemac53** [comment]
> Nobody made an issue about it. But looking back at it, I know that was a mistake.

For build to source code generators the problem is less pronounced, because you ship the generated code with your package. So the output is fixed at a point in time where it was valid, and you don't have this "transitive deps got randomly broken" issue. I think that is why you probably didn't get issues filed, the fix was always easy (and local, just update your sdk constraint).

For macros it is much more likely people would get broken in such a way that they can't fix the problem themselves, and will file issues.

**Comment by natebosch** [comment]
If we were able to guarantee that we'd have a reliable `dart migrate` for future language versions then a macro targeting an old language version could keep working during whatever support window we choose. Using the macro language version as a min constraint for usage should also work in most cases, we can maybe have an override mechanism if a use case demands it.

**Comment by jakemac53** [comment]
> If we were able to guarantee that we'd have a reliable `dart migrate` for future language versions then a macro targeting an old language version could keep working during whatever support window we choose.

Can you elaborate more on this?

**Comment by natebosch** [comment]
It might break our performance targets, but if a macro outputs code at language version `3.3` and the library is `3.4` then we could take the output and migrate it automatically to `3.4` (I assume a library and it's augmentations must have a consistent language version). So old language macro can be compatible with new language usage.

The reverse situation (and main topic of the issue) would be handled by changing the pub solver so new language macro is never used with old language usage.

**Comment by jakemac53** [comment]
> It might break our performance targets, but if a macro outputs code at language version `3.3` and the library is `3.4` then we could take the output and migrate it automatically to `3.4` (I assume a library and it's augmentations must have a consistent language version). So old language macro can be compatible with new language usage.

Ah ok, that is an interesting point, but `dart migrate` probably requires the analyzer which would make it not really feasible.

It also is less of a concern generally, given the fix is local :).

> The reverse situation (and main topic of the issue) would be handled by changing the pub solver so new language macro is never used with old language usage.

Yeah this was the solution @rrousselGit suggested as well. We would have to know that a package has any macros at all, ideally by only looking at the pubspec, and we weren't originally planning on requiring anything special in the pubspec for macros.

**Comment by natebosch** [comment]
> and we weren't originally planning on requiring anything special in the pubspec for macros. But it is probably the simplest solution.

Yeah, I think it's also something users may want to see on the pub site.

**Comment by lrhn** [comment]
> solved by the version solver

I don't think we should assume that that is possible.

If a macro can only be applied to Dart 3.0+ libraries, because the code it generates uses records, then using the version solver to prevent misuse means that the macro cannot be used by any package which has a min-SDK below **4.0**.
It's effectively unusable.

Any package with a min-SDK below 4.0 can be used by another package which has a min-SDK of 3.X, and which has a library with a `@dart=2.19`. And then the macro can be applied to that library.
(Assuming we support version 2.12-2.19 in all 3.X SDKs. Otherwise use the version where we discontinue pre-3.0 langauge versions.)

Version solving doesn't work for that scenario
The macro *must* be able to enforce that it's only *applied* to libraries that satisfy some lower (and maybe upper) bound on what it can generate code for.
The macro is unusable if it cannot be used _in the same program_ as code that it's incompatbile with, which is what version solving is about. And at that point, it has to check each application for whether it's in a supported library or not.

The min-SDK of a macro is the SDK needed to *run* the macro. (If it can only generate code for a higher language version than that, it should probably up its min-SDK too.)

The "min-supported-library-language-version" of a macro is the minimum language version it can be applied to. That should be checked for each macro application.

We can, and probably should, make that part of the macro API, something the macro must declare, and that the analyzer can check when it sees a `@MyMacro(...)`, without needing to run the macro.

Then the macro should still be able to check the language version it's generating code for, because it might want to generate *different* code for different language versions.

The upper-bound is more tricky. When we add new language features in the future, that changes the introspection API, and possibly the code-generation API. The macro API, which is used both for reading and writing, and is therefore, generally, language version *invariant*. 
A macro may not be ready for changes to that. It may see declarations it doesn't understand, which it can possibly handle gracefully, but it may still have to give up on creating code.

We can try to promise some kind of backwards compatibility, so you can always generate old code. But it won't be allowed to hold back language evolution, so we might as well be prepared for how to handle breaking changes.

We could have a "macro API version", so that any change (or any breaking change, for some definition) to the macro APIs increases the version, and a macro must declare itself compatible with the new version before it gets access to it, maybe even allow it to declare a specific entry point *per macro API version*, so that we can provide completely different types if needed.

We can then provide the newest API that a macro supports, for it to use, also for older langauge versions if they are compatible. If not, we'll have to provide the old API version for those libraries.

**Comment by natebosch** [comment]
All that said, it's likely that adding these type of backwards constraints would be challenging or impossible in the pub solver... @jonasfj

**Comment by rrousselGit** [comment]
@lrhn IMO the case of `// @dart=2.19` is separate.
In my eyes, that's akin to a dependency_overrides, and should be treated as a "the user knows what they are doing". It's also a niche use-case.
If we can catch a mistake early in that case, that's perfect. But if not, I don't think that's a big concern. 

> Then the macro should still be able to check the language version it's generating code for, because it might want to generate different code for different language versions.

I think it would be more reasonable to flip the logic.
A macro could specify `sdk: ">=2.15.0 <4.0.0`. Then, when it detects that it's run against a 3.0.0 library, generates a sealed class.

That would enable the macro to run on older language versions. As opposed to the other way around, where even though the macro supports `// @dart=x.y`, the macro would still require an up-to-date SDK locally installed.

**Comment by jakemac53** [comment]
> If a macro can only be applied to Dart 3.0+ libraries, because the code it generates uses records, then using the version solver to prevent misuse means that the macro cannot be used by any package which has a min-SDK below 4.0.
It's effectively unusable.

I don't understand why this wouldn't be possible, can you explain?

> Any package with a min-SDK below 4.0 can be used by another package which has a min-SDK of 3.X, and which has a library with a @dart=2.19. And then the macro can be applied to that library.

I agree that language version comments complicate things. I am not sure if you can alleviate this with lints or something. Ultimately, I don't think these are very common any more, and we don't _have_ to solve for that case. In other words, I would consider this a "don't let perfect be the enemy of good" scenario. Having some level of checking in the pub solver is better than having none.

Possibly we could lint for this situation or something.

> The "min-supported-library-language-version" of a macro is the minimum language version it can be applied to. That should be checked for each macro application.

I agree this could work if macros always do breaking changes when they update this version. I would just like to avoid that solution if possible.

> The upper-bound is more tricky. When we add new language features in the future, that changes the introspection API, and possibly the code-generation API. The macro API, which is used both for reading and writing, and is therefore, generally, language version _invariant_.

I agree but I am also not nearly as concerned about the _reading_ part. The difference is the kind of breakage that happens. The "reading" scenario is (mostly) going to occur when a user explicitly opts their own code into a newer language version. At that point, it is possible that some macro might fail, but the user can just reset their language version back to what it was (or possibly upgrade the macro package).

We also are planning on using a package to expose the macro apis (sort of similar to how dart_internal works). We could do breaking changes in that package when new features are released, if we thought it was useful. Which is related to the last point:

> We could have a "macro API version", so that any change (or any breaking change, for some definition) to the macro APIs increases the version, and a macro must declare itself compatible with the new version before it gets access to it, maybe even allow it to declare a specific entry point per macro API version, so that we can provide completely different types if needed. (Maybe it just needs to implement MacroV4 to be compatible with the macro V4 API, which gives it entry points named something-with-4 that can be called with a MacroApiV4 object. Or something.)

We are planning to use a package for this versioning. It will only directly export some internal library, but it will provide a way to version the macro API separately from SDK versions. This package will have tight SDK constraints.

**Comment by davidmorgan** [comment]
A few thoughts.

Generally speaking macro implementations should not be super excited about using new language features for _implementation_, by which I mean, concealed in the generated code.

Let's take the addition of records to the language as an example.

As a macro author I don't think I would update my existing macro implementation to generate _implementation_ code that uses records--I know it would make the macro apply to less code, so there would have to be a big benefit, and there likely isn't one. (Switch to records for better memory/performance?)

On the flip side if a macro is applied to code that uses records in the API there is a good chance the macro generated code also wants records in the API. So for example if a field in a `built_value` class is a record then it's likely the generated code needs records too.

Then this comes almost for free: it takes input code with records to trigger output code with records, so the language version automatically is high enough.

Stretching a bit it's possible to imagine a case where this breaks: what if the macro is applied to a file with too low language version for records, but that file references source with a higher language version that does use records, _and_ it references it in a way that triggers the macro to want to generate with records.

But is that something we'll see in practice? In the examples I've considered so far it looks like not: the main information a macro is going to want to pull from distant sources is about which types have that same macro applied, so it can assume the features it itself generates.

It would be useful if we can find concrete examples.

The most important concrete example we have so far is https://github.com/dart-lang/language/issues/3466#issuecomment-1808736529 about lints: lints that force generated code to move to new features are a big problem :) ... if we don't do that--if we instead encourage generated code to stick to the minimum language version possible ... then what kind of problems are we left with?

Thanks.

**Comment by jakemac53** [comment]
I do generally agree that if macros just stay somewhat off of the bleeding edge of the language that likely avoids many of the problems. My main concern is communicating that to macro authors in a way that is reliable and doesn't require them to break their users before understanding the issue.

**Comment by rrousselGit** [comment]
I think the main issue about staying somewhat off of the bleeding edge would be that a macro written with Dart 2.x (assuming we could write macros with it) wouldn't have access to the Ast/Elements related to 3.0 features.

So if that macro were to be applied on a sealed class, the macro couldn't do something special for it without bumping the minimum necessary Dart version.

**Comment by jakemac53** [comment]
A macro can _usually_ still support newer language versions without dropping support for older ones. It can increase its min SDK constraint (or macro API package constraint) and still be applied to packages on older language versions.

Typically it will only need to actually emit code using newer language features if they appear in the library they are running against. Consider for instance extension types, they can support running on extension types (and generating augmentations of them) without dropping support for libraries that existed prior to extension types, since those libraries won't contain any extension types.

Similarly it can support generating record types in APIs, because they will generally appear in the library they are generating code into as well. This breaks down for super types from other libraries at some point, but I don't see any way around that.

**Comment by davidmorgan** [comment]
I guess we will at some point have a recommended way for macro authors to test their implementations.

Do we have the capability anywhere to detect language version? i.e. can we give macro authors a method:

```
final macroOutput = applyMyMacro(myTestCase);
expect(macroOutput.detectedMinLanguageVersion, isLessThanOrEqualToVersion('`3.1`));
```

So that if they change the macro to output a 3.2 feature in this case, the test fails?

That would be for unit testing.

I guess for e2e testing they can easily do this with a test case input containing `// @dart=3.1`, then the test will fail if the macro outputs code >3.1. So maybe our "how to write and test macros" doc has example e2e tests in which every test case is tagged with a language version.

**Comment by rrousselGit** [comment]
@jakemac53 The problem is, if a macro relies on a higher SDK to generate code for older language levels, folks may not have the new SDK installed.

For instance, a macro may want to support both the stable and master channels of Flutter.
But if the macro bumps its SDK to change its generation based on the new features for the master channel, then the stable channel won't work anymore.

**Comment by lrhn** [comment]
> We are planning to use a package for this versioning. It will only directly export some internal library, but it will provide a way to version the macro API separately from SDK versions. This package will have tight SDK constraints.

I guess that means that each major version of that package will have a tight SDK range, and if we release a new SDK which it is compatible with, we'll release a minor version update with the wider SDK range. If not, it's a major version increment of the package.

That still means that the macro cider itself will either inherit the limited SDK range, no matter what range it will itself claim compatibility with, or it will have a package dependency which spans multiple major versions.

**Comment by davidmorgan** [comment]
Can we usefully introduce flexibility at the serialization boundary?

It would be possible to have a macro built against a newer macros API talk to an older SDK, if the serialization is forwards/backwards compatible. The older SDK is guaranteed to not send data that the macro won't understand.

The other way around is slightly more awkward, a macro built against an older macros API talking to a newer SDK. But we can have a well-defined concept of "unknown to this macros API version".

**Comment by lrhn** [comment]
> The problem is, if a macro relies on a higher SDK to generate code for older language levels, folks may not have the new SDK installed.

That's just normal dependency solving.
If your macro package requires SDK 3.7, it won't be part of a package contravariant solution run on SDK 3.6.

_Running_ the macro code is the least of the issues here, that's just a dependency any other, a late-applied dev-dependency.

The possible problems are:
- compatibility with the macro runtime API, which may change between minor SDK versions. Maybe only incrementally, but that can still be "breaking" if you assume a switch over possible to level declaration types is exhaustive. A new, unknown type of object can be surprising. If there is any kind of visitor pattern, it'll likely be just plain breaking.
- generated code compatibility with the library it's generated into. Most libraries will have a language version below the most recent SDK version, because most packages don't need to be updated on each minor SDK release. They'll happily continue with a `^3.4.0` SDK constraint until they need to change it. If the 3.5-compatible version of the macro package has the same major version as the 3.4-compatible release, and it blindly generates 3.5 Dart code, there is nothing in the version solving which can prevent this. Dart 3.5 is 3.4 compatible, in that it accepts 3.4 libraries. The macro package is SDK 3.5 compatible. Everything is fine from a versioning standpoint, it's the macro package which must generate 3.4 code, or it must make the 3.5 version a major version increment, that users must opt into, after making the libraries using them macro and 3.5 libraries.
- integration issues. If the macro is not compatible with older language versions,  and does a major version increment, then two independent packages that both use the same macro package, may end up depending on incompatible versions. Macros are dependencies, common macros will likely be shared dependencies, which make it very hard to do a major version increment. But if that's what it takes to not break at compile time, that's what will happen.

I think my conclusion is that successful macro packages will support generating code for any language version that is supported by its min-SDK, at least any version that it has supported at any time. A new macro package release will keep the major version number, and support more language versions, backwards compatibly, but never fewer. Not until the SDK itself stops supporting that language version.

**Comment by rrousselGit** [comment]
> That's just normal dependency solving.


Yes. It's just that I don't see the idea of "have a macro use `>=3.0.0`, but support running on `>=2.0.0` libraries" as sustainable due to `pub get` issues, and that we should update package resolution for macros.

From there my point was to reverse the coupling when macros want to support a range of Dart version, such that a macro would specify `>=2.0.0` and yet fully support `3.0.0` too (including generating sealed classes when need be).
But that means the macro wouldn't be able to know about the new 3.0.0 features (but likely would still be able to generate code using it, just not parse it).

So a macro cannot both:
- support running on Dart 2.0 without having to locally install Dart 3.0
- support being applied on `@macro sealed class` and do something special with sealed classes.

Meaning a macro would likely have to have two versions, one for Dart 2.0 and one for Dart 3.0.
But that would cause versioning issue too. A library may want to use Dart 2.0, but depend on a package which uses the same macro and expect it to work with 3.0.

**Comment by davidmorgan** [comment]
> * support running on Dart 2.0 without having to locally install Dart 3.0
> * support being applied on `@macro sealed class` and do something special with sealed classes.

A macro doesn't need the 3.0 SDK to understand 3.0 code, it just needs the 3.0-corresponding macros API.

Right now the macros API _does_ have a tight SDK dependency, so the 3.0-corresponding macros API does force the 3.0 SDK, but I don't think that necessarily has to be the case. As far as I can see the first barrier to loosening that up is serialization compatability, I'm not sure if there are others.

**Comment by lrhn** [comment]
The macro implementation code will be run by the current SDK while compiling a macro-use, so the macro package's min-SDK should ensure that it's *compatible* with the current SDK. The code it generates will be compiled with the same SDK, so it should not generate code for a language version above the min-SDK. 
The min-SDK of a macro package is essentially "the latest language version it knows of". Any later language version is *unknown*, and can have any amount of changes to the language, up to and including a *complete change of the syntax*. (We probably *won't* do that, but with language versioning and migration, *we can*. As long as we still support compiling earlier language versions too, it's technically non-breaking.)

> Yes. It's just that I don't see the idea of "have a macro use `>=3.0.0`, but support running on `>=2.0.0` libraries" as sustainable due to pub get issues, and that we should update package resolution for macros.

I think supporting libraries at a previous language level is essential, and the only way macros are going to be actually usable if we won't accept more than one version of the same macro-package in one program.

*Most libraries are not at the max SDK language version.* 
Because they don't need to until they want to use a new language feature. Being 3.0 is perfectly fine, and going to 3.2 for no reason, just risks inference changes that you'll have to fix. There is no rush.

If you mean a lower *major* language version, then I still think it'll be necessary. The change from 2.19 to 3.0 was no bigger than the change from 2.11 to 2.12. It's not about version numbers, it's *possibly* about breaking changes, but in the end, that still doesn't matter. A useful macro package must support all the language versions where it has ever worked.

We can forget `2.0`, I don't expect macros to be usable in any library with a language version below 3.X, where we release the macro feature. Let's, optimisitically, call it 3.4 for now.
It's not the past before that which worries me, but the future, when that becomes past.

If a single release of a macro can't support both language version 3.4 and language version 3.5 libraries, then it can't be used in the same program, because there *will* be programs having both language versions, and both using the same macro package.

**Example**

Let's say I release the tremendously successful `dataclass` macro, and everybody starts using it. It gives you `==`, `hashCode`, optionally `toJson`/`fromJson`, and a nice `toString` for debugging. What's not to like?
And within weeks, 98% of top-1000 packages use my macro package. Roaring success, macros FTW.

The `dataclass` package has version 1.0, has a min-SDK of 3.4.0, and depends on `package:macro_api` version `^1.0.0`, which requires SDK `^3.4.0` (might add `<3.5.0`, but let's assume nobody does that). That's where everything starts, where it has to start.

Then Dart SDK 3.5 is released with just some performance fixes, no language changes.
The `macro_api` package needs no changes, or few and backwards compatible changes, and it's
re-released as `1.1.0` with an SDK range of `>=3.4.0 <3.6.0`.

My macro package keeps working, it'll likely get `macro_api` version 1.1.0 when compiled on a 3.6 SDK, but that's fine. It's compatible, the version number says so, and things work. If I do any kind of language version checking on the libraries I update, I might need to change a `3.4` to `3.5` somewhere and release a patch or minor version, but everything just keeps working.

Then Dart SDK 3.6 is released with new fancy language features called, say, "record classes" and "static extension methods", or something. 

Now, either we can release `macro_api` version 1.1.0 which is backwards compatible and has an SDK range of `>=3.4.0 <=3.7.0`, or it cannot be compatible, and the version will be 2.0.0 with an SDK range of `>=3.6.0 <=3.7.0`. Let's say it has to be incompatible, and is version 2.0.0.

I want to support generating a `toJson` and a (static extension) `fromJson` for record classes,
so I release a new version of the the `dataclass` macro package which depends on SDK 3.6.0 and `macro_api` 2.0.0, which does that.

Here I have two choices. 
* Either I keep supporting 3.4 and 3.5 libraries, and also support 3.6 libraries, with some features only working (and applying) in 3.6 libraries. Maybe the `fromJson` needs to be written differently everywhere to account for "record classes", but that's fine. The code is running on a 3.6 SDK, even if the 3.4 library can't itself create a record class. The new release of the `dataclass` package will be a minor version increment. If your SDK is compatible with it, it's backwards compatible with the earlier 1.X versions.

* Or I don't, the macro now only works on language version 3.6 libraries, and I have to give the macro package a major version increment (because it's not compatible with version 1.0.0, so it should not be matched by a constraint of `^1.0.0`).

Should I choose the latter, then we have version hell.

You cannot have two versions of the same package in the same program, including macro packages.

You cannot use `dataclass` 1.1.0 on a 3.6 SDK, because it depends on `macro_api` ^1.0.0 which have a max SDK of `<3.6.0` on every version. 
That is, unless we *also* release `macro_api` 1.2.0 which supports 3.6 SDKs a at least up to language 3.5 libraries, maybe even 3.6 libraries without giving access to record classes or static extensions.

And you still cannot use it in the same library as `dataclass` 2.0, which is needed to use the macro on a record class.

So if you want to upgrade to 3.6, and use a record class with a `dataclass` macro, then there is no way you can do that until *all* your dependencies have upgraded too. Every one of them, dev-dependencies included. If `test` depends on something which still uses `dataclass` 1.1, you can't migrate.
(And unless we release `macro_api` 1.2.0, you can't even switch to using SDK 3.6.)

Popularity is a problem for versioning, when you have a one-version policy. Breaking changes are near impossible to ship, and get adapted.


**Conclusion?**

Not sure what the conclusions are:
* Never make a major version increment of `macro_api`?
* Never make a major version increment of `macro_api` without also releasing earlier versions that are "compatible" with the new SDK, at least enough to work on libraries with earlier language versions. Preferably also new language versions that don't use the new language features, or don't use macros on them.
* Never make a major version increment of a macro pacakge.
* Always support every SDK-supported language version with your macro package.

Personally I'd go for the last two, together. That's what would bring my users the most stable experience, and it's what I can control, as a package author.

Which means releasing a new version every time the `macro_api` changes major version, a release which is backwards compatible *and* supports the new language features. 
Or just backwards compatible, so the macro package keeps working on new SDKs, then I can add new feature support in separate releases. Just never leave your users stranded in version hell, where no version of your macro package can support every package depending on it.

**Comment by lrhn** [comment]
> A macro doesn't need the 3.0 SDK to understand 3.0 code

Macros do not run in a void. They are most likely/commonly being run *during* compilation, so it doesn't matter whether the macro needs the 3.X SDK to work on a 3.X library, it won't get *run* on such a library by anything less than a 3.X SDK, because the parser would have given up long before.

I think it's perfectly reasonable that a macro working on 3.X code requires a 3.X SDK. If you have a lower SDK, you need a different version of the macro package, for a lower SDK, but that won't be a big loss, since there can't be any 3.X code in the program then.

**Comment by rrousselGit** [comment]
> * Never make a major version increment of a macro pacakge.
> * Always support every SDK-supported language version with your macro package.

As a maintainer of multiple code-generators with 99-100% popularity score, this sound very unrealistic to me.  
This puts an immense burden on macro authors. Maybe one or two macros will try, but there's no way the community at large will do so.

I don't understand why it would be desirable for macro authors to support the weight of the entire world, when a flag in a `pubspec.yaml` to have libraries using macros to respect the macros' min SDK would solve most of the problem.

**Comment by davidmorgan** [comment]
> > A macro doesn't need the 3.0 SDK to understand 3.0 code
> 
> Macros do not run in a void. They are most likely/commonly being run _during_ compilation, so it doesn't matter whether the macro needs the 3.X SDK to work on a 3.X library, it won't get _run_ on such a library by anything less than a 3.X SDK, because the parser would have given up long before.

I'm not thinking of this benefiting the 3.X users, but the 2.X users. If the macro that can now generate 3.X code does not simultaneously now require a 3.X SDK, then 2.X users can get updates to the macro that are released after/alongside 3.X-only features.

> I think it's perfectly reasonable that a macro working on 3.X code requires a 3.X SDK. If you have a lower SDK, you need a different version of the macro package, for a lower SDK, but that won't be a big loss, since there can't be any 3.X code in the program then.

It is possible I am misunderstanding something, but as a macro (well, so far: codegen) maintainer this seems like a situation I want to avoid: it means that I have to backport new features unrelated to 3.X onto a 2.X branch and release them from there in order to support 2.X users, or 2.X users simply don't get them.

It's possible that "macro authors maintain a branch for N language versions and backport every new feature onto them" is part of the best practices, it's also possible it doesn't come up often.

For what it's worth, I never bothered backporting features for codegen; it was simply always the case that you had to be on a sufficiently recent analyzer to get the latest version. But this always felt like this very tight coupling was a downside/weakness; and because codegen has been distributed by checking in generated files, this restriction did not propagate, so we have not seen the potential full downsides of such pinning.

Thanks.

**Comment by natebosch** [comment]
> it means that I have to backport new features unrelated to 3.X onto a 2.X branch and release them from there in order to support 2.X users, or 2.X users simply don't get them.

Yes, users who do not upgrade don't get new features. I don't understand why it is a critical feature for a project to be able to use new features of `built_value` while retaining a stale SDK. We don't expect new language versions to be as difficult to migrate to as null safety was - there should be little reason to hold back to an old SDK.

**Comment by lrhn** [comment]
Staying on an older SDK should always be an option, then you can't expect to use any newer package releases, but the old versions should still work.

I only worry if a user of a new SDK can somehow get locked into using an older version of a package anyway, or if switching to a new SDK will make an existing package stop working (for no really good reason).

If a macro package puts a tighter upper bound on their SDK, directly or by depending on a particular macro-API package with a narrow SDK range, then it may not work on the next minor-version-increment SDK version.
That would be bad.

So, preferably, every minor SDK version increment should be backwards compatible with all macro packages written for the previous SDK version, with at most a `pub upgrade`.

Getting the benefits of the new SDK features, and supporting them with macros, should be possible in a minor version increment of a macro package.

How we achieve that doesn't really matter, as long as it's not too onerous.

**Comment by davidmorgan** [comment]
> This also alleviates any serialization format concerns, because the serialization libraries are always identical across the SDK tools and the user packages (it is always the current SDK version of the library).

My point about serialization is not about it being a concern but rather a potential benefit--serialization brings possibilities you can't get any other way.

Macros are about transforming a complex data structure into code, and the schema of that data structure can change with each new language version.

Well, there is a lot of prior art in that area :) it's what databases and protos do. The way you deal with a schema that can change is that you serialize the data, because your classes-and-static-checking-focused code can't cope with schema changes, but your serializers can.

Because of serialization there is no reason for macros to be tied to a particular SDK version--they don't even have to be written in Dart. As an extreme example, schema discovery could be made part of the protocol, so a macro would not have to know the details of the schema, even.

Why does any of this matter? Maybe it doesn't. But as a codegen author it's frustrating to have to do a version-narrowing release of

```
if (element.hasExcitingNewFeature) emitForExcitingNewFeature();
```

when a very simple API change would make it non-breaking

```
if (element.hasFeature('ExcitingNew')) emitForExcitingNewFeature();
```

--the narrower release isn't because of any choice _I_ made, it's because of an API that I'm forced to use. My code should work for both versions, all of the pieces are there, but because the data model changed _and I'm forced to compile against it_, that's not possible.

Macros seems like a chance to fix that, and the fact that there is a serialization step means we are already doing a lot of the work needed to make data model changes not quite so breaking. Serialization exactly means that code compiled against different versions of the data model can interoperate. So, a macro compiled against a _new_ macros API can work on an old SDK--for sure, because there can be no unexpected changes in the old SDK; a macro compiled against an _old_ macros API can work on a new SDK--maybe, maybe not, because the new SDK might come with data changes that break the meaning of the data model even allowing for ignoring "unknown" new parts of the data model.

> Yes, users who do not upgrade don't get new features. I don't understand why it is a critical property for a project to be able to use new features of built_value while retaining a stale SDK. We don't expect new language versions to be as difficult to migrate to as null safety was - there should be little reason to hold back to an old SDK.

This is true until it isn't :) ... a lot of `built_value` releases are bug fixes for some obscure case, so it would come down to what version of the SDK the person who needs the fix is on.

If we can address the whole issue at a high level, removing unnecessary version pinning, that seems like a win.

The benefits work the other way, too: if every language change causing a schema change causes work for all macro implementors and the ecosystem, that will be an incentive for the Dart team to avoid releasing tiny language changes, to bundle them up into bigger releases. But small releases, generally speaking, are better for release health / system health. So that's an unwanted tension.

I realize that I have wandered into territory that is off topic for the issue, sorry about that, I guess there is no need to dig into this further for now, I will open a different issue if I ever feel the need to do that :) thanks.

**Comment by lrhn** [comment]
@davidmorgan  Those are good points. My choice of solution would not go towards "serialization", which I really hear as "dynamically structured data" (like protobuf, which survives by being always backwards compatible at the API level, but then you have to check whether each property is there).

If we have SDK/compiler *integration*, and a `package:macros` API which integrates with that and exposes user-facing data, then there is a number of ways we can make that evolvable and still backwards compatible.

You shouldn't need a narrow SDK band for a new feature, just a min-SDK which is the release of that feature. That should be both nevessary and sufficient. Then you should be able to depend on  the `macros` API package corresponding to that SDK feature, and you should be OK with getting a *later* (minor) version too.

Meaning that the `macros` package for a new minor SDK version should be a new minor package release, and be backwards API compatbile with earlier versions.
Maybe the `macros` package will be bound tightly to a single SDK release, that's OK as long as we can always *solve* to get that, if everybody uses `macros: ^3.5.0` to get the version supporting the 3.5 language features.

How to then *make* it backwards compatible is a design issue. We have multiple choices.

The simplest is to have a `package:macros/3.5.dart` library for each SDK release that supports macros, and keep the old ones until we stop supporting that language version. Many will likely be identical, exports of the same underlying library. Some will have additions. Some will have breaking changes, which is why it's great that you have to opt-in to it.

If you're using `macros/3.5.dart` and running on a 3.7 SDK, the code inspection API might see declarations it's not ready for. The `3.5.dart` API can export an `UnknownTypeDeclaration` type, and similar for other syntax categories. If it sees a post-3.5 feature declaration, it can always expose it as `UnknownTypeDeclaration`. (Heck, it might be the same object, it's just that `RecordClassDeclaration` from the `3.7.dart` library implements `UnknownTypeDeclaration` from the `3.5.dart` library, as long as it's compatble, which might just mean `String get name` and `UnknownType asType()` members.)
So, generally, any hierarchy of typed, even made sealed, will contain an `UnknownFoo` or `FutureFoo` entry that inspecting later code can return.

Another option is that "you can ask" feature of `if (library.hasFeature('record-class')) ...`.
That's harder to do well in Dart, because you can't get to a good typed API for record classes without being able to import a library that declares it, which you can't if it isn't there.
(I don't believe we'll allow compiling a deferred import with a target which isn't there.)
And an entirely dynamic API is just not good usability, it has no type checking. It'll be seconds before someone writes a "complete" typed API on top, as a separate package. That might as well be us, which means you should just get the newest `macros` package. As long as it's backwards compatible, any other `macros`-package depending package will also be satisfied by it.

We might not need explicit versioning, if we can make the API changes incremental and non-breaking.
But adding a subtype to a sealed hierarchy is breaking, and we might want to do that when we add new kinds of syntax.

**Comment by rrousselGit** [comment]
> The simplest is to have a package:macros/3.5.dart 

> Another option is that "you can ask" feature of if (library.hasFeature('record-class')) ....

I don't think these are simple solutions at all.
They involve a lot of work to maintain because they require maintaining multiple implementations of the same feature based on the language version.

Even with a simple feature such as constructor tear-offs, very few generators take the extra cost of generating `fn(constructor.new)` vs `fn(() => constructor())` based on the library version they are applied to.

I would expect most macros to just always fail if the library they are applied on doesn't have access to the new language feature they need, Be it with a compilation error or an explicit macro error.

**Comment by davidmorgan** [comment]
> I don't think these are simple solutions at all. They involve a lot of work to maintain because they require maintaining multiple implementations of the same feature based on the language version.
> 
> Even with a simple feature such as constructor tear-offs, very few generators take the extra cost of generating `fn(constructor.new)` vs `fn(() => constructor())` based on the library version they are applied to.

For this example, is a good path not just: 1) generate only the old version, which works everywhere, 2) have an end to end test pinned to an old language version, to guarantee not accidentally using the new version.

Generated code does tend to get stale over time--but as long as it's implementation, i.e. hidden, that seems like a feature not a bug? :)

> And an entirely dynamic API is just not good usability, it has no type checking. It'll be seconds before someone writes a "complete" typed API on top, as a separate package. That might as well be us, which means you should just get the newest macros package. As long as it's backwards compatible, any other macros-package depending package will also be satisfied by it.

Thanks! Yes, that fits how I'm thinking about it. Filed an issue for this topic: #3479 "Macros API stability across language versions"

**Comment by rrousselGit** [comment]
> For this example, is a good path not just: 1) generate only the old version, which works everywhere, 2) have an end to end test pinned to an old language version, to guarantee not accidentally using the new version.

That assumes you don't need the new language feature for something.

Generators often receive new features which specifically require new language features. Supporting old language versions therefore can be much more complex than simply sticking to generated old code.

Let's not also forget that some macros aren't standalone. Take riverpod_generator or Mobx. Both would be migrated to macros, but both are also part of a broader SDK. And a new feature of that SDK would involve likely tweaking the generated code, which may require those new language features.

@lrhn's approach would therefore require those macros to make a major version bump whenever they hit those cases, which is undesired.

On the flip side, either having `pub get` resolution be contravariant or have two separate constraints in the pubspec of macros would avoid this:

```dart
name: macro_package
environment:
  sdk:
    # the macro can run using Dart 3.0.0
    source: ">=3.0.0"
    # The code generated by macros in this package require Dart 3.5.0
    applied: ">=3.5.0"
```


(let's ignore the `applied`/`source` name. They keywords can change)

When doing so, a macro no-longer would have to make a breaking change if a new feature bumps the minimum required SDK. The version solver would handle everything.

Folks who want to support old versions the way @lrhn proposes could still do so. But not everyone **has** to.

**Comment by rrousselGit** [comment]
A bonus of having two constraints in the pubspec.yaml is that it could also spot constraint mistakes.

If a macro wants to support `>=2.0.0` (random number, I know Dart 2.0.0 won't have access to macros) but mistakenly generates a record, the macro's test suite would naturally catch this without any fancy testing setup.

Instead of having a CI matrix running a test suite against multiple environments, we could do that macros naturally generate a `// @dart=<minimum macro's supported dart version specified in the pubspec.yaml>`.

Then the test suite would automatically fail with a compilation error if generated code contains features that do not respect the constraints.

**Comment by sigurdm** [comment]
Any chance that the macro-generated code could always be interpreted in the language-version of the macro itself instead of its invocation context? Would that even make sense?

**Comment by lrhn** [comment]
> Any chance that the macro-generated code could always be interpreted in the language-version of the macro itself instead of its invocation context? Would that even make sense?

Maybe!
For some language version differences, it's effectively the same as just upgrading the library to the macro's version, when all changes in-between are incremental.
If not, then it's still workable if all the changes are inside function bodies, because each body is entirely in one file.

Where it gets tricky is changes like class modifiers, where declarations themselves hers new functionality and restrictions. If a macro adds a `base` modifier to a class which is seemingly declared in a pre-feature library, what does it mean?
Maybe we can figure that out, but it's not obvious that it'll always be possible.

This is mainly an issue because augmentations are inside the same library, and while we define the interaction of features with pre-feature libraries, that mostly assumes that each library has a consistent semantics. We don't have to explain how an instance member from one language version interacts with a surrounding class declaration with a different language version.

**Comment by natebosch** [comment]
> This is true until it isn't :) ... a lot of `built_value` releases are bug fixes for some obscure case, so it would come down to what version of the SDK the person who needs the fix is on.

How often has this come up where a user needs a fix and refuses to, or would be unduly burdened if they needed to, use the latest SDK? Are there projects that take the approach of holding back SDK updates (out of fear of change causing instability?) but _don't_ hold back package updates (out of the assumption that pub packages are more stable than the SDK?).

How do you decided how wide this window of support should be? How old of an SDK should we try to support for projects which continue to want to upgrade only their package dependencies without upgrading their SDK dependency?

**Comment by sigurdm** [comment]
> Maybe we can figure that out, but it's not obvious that it'll always be possible.

If at all possible to define semantics here, I think it would be much preferable to putting the burden on the macro to know about what language version it generates for.

**Comment by davidmorgan** [comment]
> > This is true until it isn't :) ... a lot of `built_value` releases are bug fixes for some obscure case, so it would come down to what version of the SDK the person who needs the fix is on.
> 
> How often has this come up where a user needs a fix and refuses to, or would be unduly burdened if they needed to, use the latest SDK? Are there projects that take the approach of holding back SDK updates (out of fear of change causing instability?) but _don't_ hold back package updates (out of the assumption that pub packages are more stable than the SDK?).

I don't know; I guess my fear of codegen and SDK versioning mostly comes from google3, which often has constraints that are very far from the norm. There certainly _could_ be large monorepos where rolling a new Dart SDK version is an expensive+slow operation, I don't know if there are :)

> How do you decided how wide this window of support should be? How old of an SDK should we try to support for projects which continue to want to upgrade only their package dependencies without upgrading their SDK dependency?

I think it's more like the guidance that goes with "don't optimize prematurely", i.e. "don't pessimize prematurely": we shouldn't throw away version compatibility if we can see a way that is technically sound and reasonably cheap to keep it.

I agree it's not worth spending too much effort on.

Thanks.

**Comment by jonasfj** [comment]
**dumb idea**: Don't allow macros to generate code using new language features until Dart 4.


**Example:** Suppose macros comes out in Dart `3.4`. In Dart `3.5` new language features are added and the introspection API in package `macros: ^3.5.0` allows macros to see the new language features.
But critically, code generated by macros can still **only** use language features from `3.4`.

After a lot of Dart 3.x releases, `3.6`, `3.7`, `3.8`, `3.9`, and so on, we decide to make Dart `4.0`.

In Dart 4.x:
 * Macros can output language features available in `3.7`.
   (coincidentally, `3.7` is where a bunch of awesome language features were added!)
 * Packages with an SDK constraint of `^3.7.0` or later have their SDK constraint reinterpreted as `>=3.7.0 <5.0.0`.
   (Meaning packages with an SDK constraint of `^3.7.0` still works in Dart 4.x)
 * Packages with an SDK constraint like `^3.6.0` or less, will not work in Dart 4.x

This is kind of how it worked with 3.x as well, everything with SDK constraint `^2.12.0` or later works. But packages with constraints like  `^2.10.0` (which aren't null-safety migrated) don't work in Dart 3.x.


--------------

Possibly, it puts some limits on how useful macros will be. And it probably requires semantics for how a 3.5 augmentation works in a 3.6 library to be defined. And possibly, I missed something else :rofl:

**Comment by davidmorgan** [comment]
@jonasfj That's an interesting idea :) it certainly would simplify things a lot.

I wonder if there is an achievable middle ground where new language features can be used in macro generated code if and only if it's not possible to generate equivalent code using earlier language versions.

**Comment by jakemac53** [comment]
> **dumb idea**: Don't allow macros to generate code using new language features until Dart 4.

The problem is they may _need_ to generate code using new language features if the library they are running on uses those language features.

**Comment by rrousselGit** [comment]
If new features aren't needed by macros until Dart 4, does Dart even need those new features? 😝

**Comment by lrhn** [comment]
> The problem is they may need to generate code using new language features if the library they are running on uses those language features.

A macro package can always refuse to work on a library with a language version above what it understands. Or it can choose to (try to) work, but only when applied to declarations that it understands. So a macro that works on classes can try to work on classes in a later language version library than what it knows about, but reject to work on a `record class` or another new and unknown declaration. 
Even attempting to work on known declarations can still fail, maybe because the known declaration contains a new unknown member (say, a local typedef), or because the generated code contains some construct which has changed meaning (say, due to "inference-update-7"), but that's not much worse than just refusing to work at all.

As long as the macro *still* works on a lower-language-level library the way it used to do, even on a new SDK.

**Comment by rrousselGit** [comment]
I think all these discussions ignore one core issue:
Most macro authors will not even know about this problem.

They will 100% make mistakes, likely multiple times.
After maintaining 4 code-generators over multiple years, I still recently made the mistake. Only then did I realize what the issue was and started this discussion.

For the sake of having a good default behavior, I think it is critical to have a good default.

Expecting certain good practices of macro authors is not a good default IMO.

**Comment by jonasfj** [comment]
> Expecting certain good practices of macro authors is not a good default IMO.

Agreed, but we do already expect package owners not to break their APIs unless they do a major version bump.

That said, I'd agree that it much easier to think it's fine to: update your macro to generate code that uses new language features.

That's not fine, and something to help users avoid doing so by accident is probably warrented. That could be warnings at publishing time, lints (maybe), different code-gen APIs for different language versions, or annotations in source or pubspec.yaml, or ?

**Comment by rrousselGit** [comment]
For me this issue is not tied to breaking change.  
The issue happens even if there's only a single version of a published package or if every version are major bumps.

A brand new user of a macro could still install the macro and end-up with a compilation error in the generated code, because their min SDK doesn't match the macro's min SDK.

That's bad experience, regardless of what happens when a new version is released.

**Comment by lrhn** [comment]
The "my min-SDK is lower than any min-SDK for my dependency" problem is not unique to macro packages. It's just easier to hit when depending on packages that increase their min-SDK or major version often.

If I have my package `foo` with SDK constraint `^3.0.0`, and I add a dependency on a completely new package `bar` which only has one version with an SDK constraint of `^3.2.0`, then I won't be able to version-solve at all on a pre-3.2 SDK, and it'll just silently work on a post-3.2 SDK.

That's not something you can see just from looking at a single package version solve. If it worked, all it proves is that there exists a solution for the _current_ SDK.
(Should `pub downgrade` behave as if the current SDK was the current package's min-SDK? Does it?)


I really do want all existing code to keep running successfully on a new minor version SDK increment.
That means that either `package:macros` (our own macro API package) must be able to solve to an earlier version with a new SDK, or (preferably) the new version we release along with the new SDK will be backwards compatible.
The latter option is preferable, because otherwise it becomes hard to upgrade any program which has two different macros in it, it won't be able to choose the new version until both macros have released a version compatible with it.

Maybe that's OK. But it's going to be annoying.

The macro code generation API should definitely warn or err if a macro tries to generate new features into pre-feature libraries. But it shouldn't prevent a macro from generating older code into older libraries, or older code into newer libraries, where it will still work. 
Only the macro itself can know what its capable of.

The macro package can always require a min SDK version big enough for it to have the features it needs.
It can refuse to work on libraries with language versions prior to that. Doing a `pub upgrade` is *not* guaranteed to be breakage free. 
The only thing that should be guaranteed is that upgrading the SDK itself, and keeping the code and all the dependencies where they are, will keep working.

**Comment by jakemac53** [comment]
> If I have my package foo with SDK constraint ^3.0.0, and I add a dependency on a completely new package bar which only has one version with an SDK constraint of ^3.2.0, then I won't be able to version-solve at all on a pre-3.2 SDK, and it'll just silently work on a post-3.2 SDK.

This is (one reason) why packages should run tests against their lower SDK bound, as well as at least one of the latest stable or dev. If you can't get a version solve, it will fail on that job, and you can increase the lower bound such that it is actually real. And you shouldn't claim support for a version you can't test.

Language versioning being tied to the min SDK does throw a bit of a wrench into this, you might end up needing to increase your language version at the same time. These things should have never been so tightly linked, but luckily most language versions are easy to migrate to.

> I really do want all existing code to keep running successfully on a new minor version SDK increment.

Absolutely, and it will, this is really the primary reason for `package:macros`. It allows us to version separately from the SDK. It is only if we do an actual breaking change in the macro APIs that packages won't be able to get a valid solve.

> The latter option is preferable, because otherwise it becomes hard to upgrade any program which has two different macros in it, it won't be able to choose the new version until both macros have released a version compatible with it.

The cost is likely to be very high though, and it isn't even clear this would be possible for all categories of breaking changes.

Note that being a package, we really don't have to address this right now. We will have the flexibility in the future to do either option (a regular breaking change, add a new library with the API changes, etc).

> The macro code generation API should definitely warn or err if a macro tries to generate new features into pre-feature libraries. But it shouldn't prevent a macro from generating older code into older libraries, or older code into newer libraries, where it will still work.

I don't think the code generation API itself needs to do this, whatever tool that is consuming the generated code is just going to treat it as normal code, and produce whatever errors it needs to if invalid code is generated. Macro generated code doesn't get any kind of back door to bypass static checking.

> The only thing that should be guaranteed is that upgrading the SDK itself, and keeping the code and all the dependencies where they are, will keep working.

*If you can still get a valid version solve*, yes. It has never been the case that we guarantee a valid version solve across SDK releases. We should certainly strive to have that be the case the vast majority of the time, but with flutter package pinning and max SDK constraints this guarantee isn't realistic.

**Comment by rrousselGit** [comment]
> The "my min-SDK is lower than any min-SDK for my dependency" problem is not unique to macro packages. It's just easier to hit when depending on packages that increase their min-SDK or major version often.

I do think it is unique to macros.  
The case you described would not result in a compilation error, but a `pub get` error:

```
The current Dart SDK version is 3.3.0-152.0.dev.

Because bar depends on foo from path which requires SDK version >=3.2.0 <4.0.0, version solving
  failed.
```

And if a version of `foo` and `bar` is compatible with the locally installed SDK and the version constraints, `pub` would use that.

Macros currently don't meet that standard. `Pub` could resolve, and instead, we would have and error when running `flutter run` or another similar command.   
That's much more cryptic, which is IMO the crux of this issue.


Afaik, the only case where there's a similar issue without macros is when using using `dependency_overrides`. But that's normal as `dependency_overrides` is about ignoring error, I expect it to have less intuitive error messages.


____


I think we're going circle. I don't want to push too much.

But there's one thing I'm failing to understand at the moment. What's the problem with adding a new optional entry in `pubspec.yaml` to ease-up this issue? (cf https://github.com/dart-lang/language/issues/3466#issuecomment-1824418154)

There's clearly some pushback against it. But I'm not sure why.

**Comment by lrhn** [comment]
>> The only thing that should be guaranteed is that upgrading the SDK itself, and keeping the code and all the dependencies where they are, will keep working.

> If you can still get a valid version solve, yes. 

I'm saying that you _must_ be able to get a version solve. Otherwise I'll consider that a breaking change for the SDK.

A new minor version of the SDK should be able to achieve the same version solve as the previous version.
The only reason to fail that is a package with a max SDK constraint which is not the next major version, which no user package should do. And no Dart team package should do unless we provide a compatible version along with the SDK, so that that failure will never happen.

That's the goal I'd aim for, and anything less than that is not a good user experience.

**Comment by jakemac53** [comment]
> I'm saying that you _must_ be able to get a version solve. Otherwise I'll consider that a breaking change for the SDK.

@leafpetersen do you want to weigh in here, we are going in circles

**Comment by leafpetersen** [comment]
> I'm saying that you _must_ be able to get a version solve. Otherwise I'll consider that a breaking change for the SDK.
> 
> A new minor version of the SDK should be able to achieve the same version solve as the previous version.

@lrhn I don't understand what you're saying here.  Whether or not it is a best practice to not have a max SDK constraint (or pin to an SDK) it is definitely possible for packages to do so.  And moreover, even if there is *a* version solve, there is absolutely no expectation that the *same* version solve will work.  We have packages that we tightly bind to the SDK: when we release an SDK, we rev those packages.  That absolutely means that an existing solve may stop working, right?

So I don't understand what you mean when you say "I'll consider that a breaking change for the SDK".  This has nothing to do with the SDK, right?

**Comment by rrousselGit** [comment]
This is a good summary!

I think this may be missing the usability concern with regards to macros being code producers.
If a macro unconditionally generated codes for >=3.5.0, even without doing anything, we would naturally always have some form of error along the way.

The problem is twofold:
- It isn't obvious that the min SDK constraint of a macro doesn't apply to code produced by said macro.
  This promotes versioning mistakes.
- Users may face a subpar behaviour. They may be able to install a macro even though is it currently incompatible with their application. This leads users into believing that there is a bug in the macro rather than an incompatible constraint between the macro and their app – thus leading to the creating of redundant bug reports on macro packages.

**Comment by lrhn** [comment]
> Require the macro author to register the language version that they expect to generate with the introspection API?

The macro author should always be able to know the language version of the library they are centering code into.
That's a fixed property of the library in the current program, decided by the applicable `package_config.json` and the location of the library file. (Possibly overridden by `//@dart=`, which we can either ignore or reflect to the macro author.)
The generated augmentation file will have the same language version as the library it's part of, so the limits of what you can safely generate is statically known.

Our macro API should be able to say, for each library, whether any known language feature is supported by it.
Say, `libraryInspector. Ok anguageFeatures.nullSafety`, which is now always `true`.

A macro author should not just generate, fx, records because they see a record somewhere. Or maybe they do, because it's necessary to satisfy the macro annotations on a pre-record-feature library which refer to a post-feature typedef from another library, and then it just fails to compile. But if someone used a macro which generates literals of some predictable type, and literals of that type are not allowed in this library, then you're holding the macro wrong.

But they should _easily_ be able to check whether the records feature is supported or not, if they want to err gracefully.

Maybe we can prevent it, if the code generation goes through an API that is versioned by the library language version, but more likely we can just make it easy to avoid creating something invalid, if one tries.

**Comment by lrhn** [comment]
> It isn't obvious that the min SDK constraint of a macro doesn't apply to code produced by said macro. This promotes versioning mistakes.

It can apply or not, but that's a property of the macro implementation.
If a macro implementation only works on libraries with language version of 3.5 or above, then it definitely won't work on a lower SDK.
Or likely won't apply to any library of a package with min-sdk 3.4.0.
It may apply to libraries of a package with min-sdk 3.6. (Unless they use version overrides, so I guess this is the place where we can ignore version overrides, in resolution, not on the actual macro.)

That doesn't mean that that particular macro can't be used in a program containing a version 3.4 package, that package will just not have a direct dependency on the macro package. And therefore not use it directly.

I guess a "minimum applicable language level" for a macro package makes sense for this. A macro can declare itself incompatible with a direct dependee that won't have any libraries the macro can apply to, ensuring that if the package did have a dependency, it'll get a prior version of the macro.

Not sure how solving will work for this.

**Comment by lrhn** [comment]
> @lrhn I don't understand what you're saying here.

It's true that we can't always guarantee that a new SDK is valid for an existing package solve. I'll bet that in most cases, it will be, because while users _can_ have SDK upper bounds, most of them really have no reason to. There is no benefit.

If we make a package which does have an SDK upper bound, I consider it our damn responsibility to ensure that there is a new backwards compatible version when the next minor version and otherwise unsupported SDK is released, so that a `dart pub get` will succeed, and the code will run.

Otherwise upgrading the SDK will break existing code that uses only `^` constraints. It might not be the SDK itself breaking things, but if it's our own widely used package that isn't compatible with the new minor version SDK, then that's only a philosophical distinction.

If `package:test` broke when we released a minor version SDK, then that would be a P0 breaking bug.
I'd expect `package:macros` to be the same.

**Comment by leafpetersen** [comment]
> The generated augmentation file will have the same language version as the library it's part of, so the limits of what you can safely generate is statically known.

@lrhn I think it would be useful for us to be very precise about what we mean by statically here, because it's one of the confusing things about macros and about this discussion.  "At runtime" for a macro is "at compile time" for a program.  Which do you mean here by statically?  More generally, I think the discussion here touches on multiple different points in time, any of which could be considered "statically known".
  - At the point that the macro is written (independently of the package it is used in or the code it is applied to).
  - At the point that a macro package is included into a pub solve (independently of the code it is applied to).
  - At the point that a macro is applied (independently of the output of the macro)
  - At the point that the output of the macro is generated (independently of the output of the compiler when compiling that code)
  - At the point that the output of the macro is compiled.

Different solutions afford different user experiences here, which do we aim for?

**Comment by munificent** [comment]
Who knew that having a rapidly evolving language, supporting multiple versions of a language in a single program, requiring packages to share versions of their dependencies, and adding static typing would lead to complex versioning problems? 😆🫠

In case another voice helps, here's mine. This thread wandered off into discussing issues around evolving the macro API itself but I'll ignore that here. The core problem in this thread is:

1. A macro may only support generating code for language version X.
2. After package resolution, even respecting every dependency's SDK constraints, a user program may end up trying to run that macro on a library whose language version is < X.
3. When that happens, the resulting code may not even compile. It may be deep in some transitive dependency that the user knows nothing about. They have no obvious way to get themselves unbroken, and it may not even be *possible* to unbreak themselves.

In practice, I think this will be a fairly rare problem. But when it *does* happen, the user will be both profoundly confused about what's going on (a compile error in some library they know nothing about) and have no clear path to un-wedging themselves. Even if uncommon, I think the latter raises the severity of this concern.

I think it makes sense to try to tackle it using pub during version solving:

*  This fundamentally *is* a version compatibility and dependency problem. You have a package that depends on some macro from another package. They must agree on a certain version (in this case, the language version of the library the macro is applied to and the language version of the code hte macro generates). Pub is the main place where we deal with version compatibility.

*  If a user gets hit by this problem, any fix they would do would involve pub. Probably tightening some constraints or adding a gross dependency_override. So if *we* don't solve it in pub somehow, they probably will on a case-by-case basis.

Here's an idea for how we could model this in the solver:

1.  Any package that has a publicly visible macro is required to add a `generated_code` version key to the pubspec. This documents what language version the macro's output uses. More concretely, it documents *the minimum language version of a library the macro can be applied to.*

2.  When version solving, if package A depends on B and B has a `generated_code` key, then we only select versions of A whose min SDK constraint is at or above the `generated_code` version. In other words, depending on a macro package that generates code at some language version implicitly raises your min SDK to that version. (This doesn't help with `@dart=` libraries, but I agree with others that we can ignore those and treat them as the user's problem).

Would that work?

**Comment by jakemac53** [comment]
Yes, that is IMO the best fix. I think it should possibly be slightly generalized, and less specific to macros, although I don't know what to call it. But essentially this is just a way of enforcing the min dart library version a package can have, in order to depend on your package.

The original suggestion was I think to just always do this _by default_ for any package with macros, using the macros own min SDK constraint as the min dart library version that is required to depend on the macro. This would also work but is slightly less flexible.

**Comment by sigurdm** [comment]
I really don't think having more kinds of versioning constraints is what we want here.

Macro's should be designed to "just work"(tm) especially given that we are only allowing augmentations to existing code, we should put an effort in defining augmentations (and thereby macros) such that different language versions can interact.

We end up compiling everything to the same internal representation anyway, so it should be possible to define what augmenting a 2.x class with a 3.y (or vice versa) macro means (we anyway are going to run 2.x and 3.y code in the same isolate). In the few cases where things cannot be well defined, I think it is ok for the compiler to give up with an error: "Sorry you cannot extend a pre-3.0 class with a newfancyfeature-method."

That is easy for me (pub developer) to say, but I think that will make the maintaining macros 100x more smooth (no need to release an update for each new language version, no need to have different codepaths for generating code for different versions, no need to understand the implications of a new kind of constraint...).

**Comment by jakemac53** [comment]
> Macro's should be designed to "just work"(tm) especially given that we are only allowing augmentations to existing code, we should put an effort in defining augmentations (and thereby macros) such that different language versions can interact.

This implies either multi-language-version _libraries_ or changing augmentations such that they aren't a part of the same library. Both of those I think are actually very challenging issues to solve, much more so than the solver solution.

But, of course that is easy to say as a language spec writer :).

> no need to release an update for each new language version

There are very few cases where I think this change would help here. It will already be the case that macros will not have to update for most language versions. As an example, many code generators today generate code into part files, and it is extremely rare for them to have to update due to a new language version. We have good existing evidence this will not be a problem.

> no need to have different codepaths for generating code for different versions

While some macros may choose to do this, it should also be exceedingly rare. Very few code generators today do this and they work out just fine.

It might be slightly more problematic than previously because the code generator (macro now) is no longer a dev dependency, but version solving should be sufficient to resolve the issue.
